utils/src/maybe_encrypted.rs

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::marker::PhantomData;
+
+use crypto::{random::make_true_rng, symkey::SymmetricKey};
+use serialization::{Decode, DecodeAll, Encode};
+use zeroize::Zeroizing;
+
+use thiserror::Error;
+
+#[derive(Debug, PartialEq, Eq, Clone, Error)]
+pub enum MaybeEncryptedError {
+    #[error("Error in decryption: {0}")]
+    DecryptionError(#[from] crypto::symkey::Error),
+
+    #[error("Error in decoding: {0}")]
+    DecodingError(#[from] serialization::Error),
+}
+
+/// A generic type that stores a value as a vector of bytes that are optionally encrypted
+#[derive(Encode, Decode)]
```

Comment:
Just to be clear: Decode is derived, which means the construction rules can be circumvented.

Besides, DecodingError isn't used. I'm assuming you're not done with this yet?

Reply:
Added a test for decoding failure. In the case of the DB I put a panic in the case of a decoding error as it means the DB has been corrupted.